src/ec-evaluator/interpreter.ts

```diff
@@ -557,69 +547,126 @@ const cmdEvaluators: { [type: string]: CmdEvaluator } = {
       control.push(instr.evalVarInstr(name, command.srcNode));
       return;
     }
-    stash.push(context.environment.getVariable(command.symbol));
+    stash.push(environment.getVariable(command.symbol));
   },
 
   [InstrType.RES]: (
     command: ResInstr,
-    context: Context,
+    environment: Environment,
     control: Control,
     stash: Stash,
   ) => {
-    // TODO throw NullPointerException if instance field but instance is null
     const varOrClass = stash.pop()! as Variable | Class;
-    console.assert(varOrClass.kind !== "Variable" || varOrClass.value.kind === "Object");
-    const v = varOrClass.kind === "Variable"
-      ? (varOrClass.value as Object).frame.getVariable(command.name)
-      : /*varOrClass.kind === "Class" ?*/ varOrClass.frame.getVariable(command.name);
-    stash.push(v);
+    
+    // Name resolution is based on the target type, not target obj.
+    let variable;
+    if (varOrClass.kind === StructType.VARIABLE) {
+      // TODO refactor? logic abit confusing
+      // Check recursively if static field.
+      let c = environment.getClass(varOrClass.type);
+      while (!c.staticFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name) 
+        && !c.instanceFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name)
+        && c.superclass) {
+        c = c.superclass;
+      }
+
+      // If static field.
+      if (c.staticFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name)) {
+        variable = c.frame.getVariable(command.name);
+      // If instance field.
+      } else if (c.instanceFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name)) {
+        // Throw NullPointerException if instance field but target is null.
+        if (isNull(varOrClass.value as Literal | Object)) {
+          throw new errors.NullPointerException();
+        }
```

Comment:
Can probably extract out 
`c.staticFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name`
and 
`c.instanceFields.find(f => f.variableDeclaratorList[0].variableDeclaratorId === command.name` 
or even put it as a method in the `staticFields` and `instanceFields` types

Reply:
> put it as a method in the staticFields and instanceFields types
Not really sure what this mean, but hope the refactored code works?